### ISSUE TITLE
collection-level user privileges

### ISSUE THREAD

**Issue by kdarras** [ecomontec/ecoSound-web#397]
We need fine-tuned access and management rights for single collections within projects (instead of all or none of the collections)
There should be a visual separation between recordings (primary data) and tags (secondary data) in the privileges window.
Checkboxes of upper levels should allow for selecting all/none of the sub-items.

Note that since sites are automatically assigned to entire projects, sites from collections for which the current user does not have management rights for should not be accessible.

**Comment by kdarras** [comment]
Column headers should have tooltips for informing users, and we could have two column heading levels to do the visual separation:

Column header "Recordings":
- subheading "access" tooltip: "access recordings for navigation, tagging, and analysis"
- subheading "manage" tooltip: "upload and delete recordings, edit their metadata in dashboard"

Column header "Tags":
- subheading "view" tooltip: "access tags of other users in player"
- subheading "review" tooltip: "review tags of other users"

**Comment by kdarras** [comment]
This seems to work globally. However:

- Users who do not have management privileges for all collections of a project (i.e., they are not project managers, only collection managers) should not be allowed to create new collections in the project that they can only access
- Users should not be able to upload recordings into collections that they don't have the management rights for
"normal user" was able to upload recordings to "test collection"
![image](https://github.com/user-attachments/assets/0e0e1a3f-110a-42b3-a614-fabde4286dba)
although he does not have the rights for it:
![image](https://github.com/user-attachments/assets/e544f562-cc0c-457d-850b-37f30f880f8a)

**Comment by kdarras** [comment]
I tested the new feature. The core works but there is a problem with project manager privileges.

I created a new project and a user with privileges to manage it. However, this new project manager cannot create new collections in that project:
![image](https://github.com/user-attachments/assets/6563ee51-23ae-4a1f-b5c2-2868e87ae80a)

Although he has management rights on the project:
![image](https://github.com/user-attachments/assets/e0d09b01-de20-45e7-a8dc-ce698b472e60)
We should thus probably add "and create collections if the user can manage the project" to the "Manage" info popup.

Furthermore, the project manager cannot give management rights to users (that he created) for single collections within his project:
![image](https://github.com/user-attachments/assets/55ea0e58-db0a-46e7-b2c9-52d056cc0f44)
However, this should be possible.

**Comment by kdarras** [comment]
We still need to add "and create collections if the user can manage the project" for clarifying this info popup:
![image](https://github.com/user-attachments/assets/64fac36c-9cdf-4bdb-93a3-31296695ef56)

To test the issue, I:

1. created a second project with the admin
2. created a collection inside that new project ("second collection for normal user")
3. created a user to manage this project (i.e. a project manager)
4. logged out of admin account
5. logged in with project manager account
6. created a new collection ("third collection not for normal user")
7. created a normal user

Then, I wanted to use the project manager account to set the privileges for the new user, but the access checkbox is disabled, and reviewing permissions were given by default:
![Screenshot from 2024-11-14 14-00-50](https://github.com/user-attachments/assets/fe89f982-41b5-4164-91fe-1db0ce9ef5c7)
Both of these should not happen.

**Comment by LiuDilongNJ** [comment]
I have added a small feature here, but it seems to be difficult to implement. When the user has a certain permission for all the collections in the project, when we add a collection, it will automatically add the same permission as the project. For example, if the user is the manage of the project, when we add a collection, it is still the manage of the new collection. If there are four collections in a project, and a user is the view authority of three collections, and the manage authority of the fourth collection, then when we add a collection, he will automatically add the view authority with the new collection. If there are four collections in a project, and a user has the manage permission for three collections and not for the fourth collection, then after adding a new collection, he will not have the permission for the new collection. But because of the initial design of the permissions function, it's id1 corresponds to view, 2 corresponds to review, 3 corresponds to access, 4 corresponds to manage. their order is confusing, we can't sort their permissions range by size, so there is an automatic increase in the id of the review permissions of 2, it should have been 1 view permissions.

So, I now have four options to choose from: 1.
1. When adding a new collection, do not change any permissions. (Doesn't seem very reasonable.)
2. When a new collection is added, add permissions to the new collection only for the project manager.
3. When adding a new collection, users with the same permissions will add permissions to the new collection. If a user has partial view permissions and partial manage permissions, they will not add view permissions to the collection based on the lowest permissions.
4. As I said at the beginning.

**Comment by kdarras** [comment]
I understand, thanks for this detailed explanation. I think that:

- admins have management rights for all projects and collections
- if a project manager creates a new collection (in his project), he and the other project managers should have management rights for it. Existing users should only get access rights to the new collection if they had access rights to the project (i.e., they had access rights for all the project's collections).
- if a project manager creates a new user, the new user should have access to all collections of the project manager's project
- users who are not project managers (i.e., they do not have management rights for all the project's collections) should have no rights for creating collections.

Sorry for writing all this; I had to wrap my head around the problem and rephrase it to understand. I think this is the same as your proposed solution.

The feature works as intended, but I added #530 to clear up the confusion around privileges.

However, for the sites, there is a problem. It looks like this normal user, who only has management rights for 1 out of 2 collections inside the "third project", can add sites to the "testing project", which is only managed by the admin:
![image](https://github.com/user-attachments/assets/86e2e63f-22d6-4b2d-acee-0301e74a24a4)

I did not manage to reproduce the weird behavior above.
This should not happen (no error message popped up). Added this: #531

**Comment by LiuDilongNJ** [comment]
I've hidden the items that should not be shown according to #531, but the tag and review functions are still shown as they can be accessed by normal users.

**Comment by kdarras** [comment]
I just created a third collection but the default permissions for the pre-existing Normal User 2 are not as expected: the user got view and review permissions although he did not have these permissions for the two preexisting collections:
![Screenshot from 2024-12-02 12-50-44](https://github.com/user-attachments/assets/e56f6228-c665-46a9-b2d0-5d61d63169bd)
Also, I cannot uncheck the gray access box for the second and third collections. I can only reset them after clicking "Manage". Inactive grey access checkboxes should be reserved for public collections.

**Comment by kdarras** [comment]
Not sure whether it is related, but it does not occur on the production server. Whenever I enter a wrong password, I get a blank http://localhost:8080/login/login page...

**Comment by kdarras** [comment]
The project manager just created a third normal user who obtained management privileges on the AAD collection even though only one of the two previous users had management rights on it. This should not happen, following our logic:
![image](https://github.com/user-attachments/assets/53c1eafb-eba0-4555-8a5e-94b212f5a427)

The default privileges system becomes quite complicated, and it will be difficult for the users to understand too... Maybe it is best to keep it simple:

- after user creation, bring up the privileges window (#530 , welcome to include in ECODIV-task2) and only pre-check the access boxes for all projects that the admin manages. I don't know the rules at the moment and I don't think anything more complex makes sense.
- after collection creation, keep the current system (users get the privilege that they had on all other collections of the project)

**Comment by kdarras** [comment]
Thanks Dilong,
Although deletion of a meta-recording by a normal user who has management privileges for a collection works after refresh, the "Error undefined" message still pops up:
![Screenshot from 2024-12-16 11-08-48](https://github.com/user-attachments/assets/6e61cad2-0e2e-4828-8cdf-a60a0c9eb3ea)
This pop-up info:
![image](https://github.com/user-attachments/assets/b6930797-925b-4841-8414-0c963b97ac35)
needs some clarification. Let's write:
"upload, edit, and delete recordings, create users and collections (for project managers)"
